lock-api/src/main/java/com/palantir/lock/client/ProfilingTimelockService.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    @Override
+    public long currentTimeMillis() {
+        return runTaskTimed("currentTimeMillis", delegate::currentTimeMillis);
+    }
+
+    private <T> T runTaskTimed(String actionName, Supplier<T> action) {
+        Stopwatch stopwatch = stopwatchSupplier.get();
+        try {
+            T result = action.get();
+            trackSuccessfulAction(actionName, stopwatch);
+            return result;
+        } catch (RuntimeException e) {
+            trackFailedAction(actionName, stopwatch, e);
+            throw e;
+        }
+    }
+
+    private void trackSuccessfulAction(String actionName, Stopwatch stopwatch) {
+        trackActionAndMaybeLog(actionName, stopwatch, Optional.empty());
+    }
+
+    private void trackFailedAction(String actionName, Stopwatch stopwatch, RuntimeException exception) {
+        trackActionAndMaybeLog(actionName, stopwatch, Optional.of(exception));
+    }
+
+    private void trackActionAndMaybeLog(String actionName, Stopwatch stopwatch, Optional<Exception> failure) {
+        stopwatch.stop();
+        accumulateSlowOperationTracking(actionName, stopwatch, failure);
+        tryFlushLogs();
+    }
+
+    private void accumulateSlowOperationTracking(String actionName, Stopwatch stopwatch, Optional<Exception> failure) {
+        if (stopwatchDescribesSlowOperation(stopwatch)) {
+            slowestOperation.accumulateAndGet(
+                    Optional.of(ImmutableActionProfile.of(actionName, stopwatch.elapsed(), failure)),
+                    (existing, update) -> {
+                        if (!existing.isPresent()) {
+                            return update;
+                        }
+                        return existing.get().duration().compareTo(stopwatch.elapsed()) < 0 ? update : existing;
+                    });
+        }
+    }
+
+    private void tryFlushLogs() {
+        if (loggingPermissionSupplier.getAsBoolean()) {
+            Optional<ActionProfile> actionProfile = slowestOperation.getAndSet(Optional.empty());
+            if (actionProfile.isPresent()) {
+                ActionProfile presentActionProfile = actionProfile.get();
+                logger.info("Call to TimeLockService#{} took {}. This was the slowest operation that completed"
+                                + " in the last {}. The operation completed with outcome {} - if it failed, the error"
+                                + " was {}. We include a stack trace of the call: {}",
+                        SafeArg.of("actionName", presentActionProfile.actionName()),
+                        SafeArg.of("duration", presentActionProfile.duration()),
+                        SafeArg.of("timeWindow", LOGGING_TIME_WINDOW),
+                        SafeArg.of("outcome", presentActionProfile.failure().isPresent() ? "fail" : "success"),
+                        UnsafeArg.of("failure", presentActionProfile.failure()),
+                        SafeArg.of("stacktrace", new RuntimeException().getStackTrace()));
```

Comment:
somewhat worried that a fast operation may be slowed down by a previous window with slow calls when they have to dump a stack trace, unclear how expensive it is.

Reply:
Hmm, yeah on second thoughts I'll remove this. If we really need to investigate just get someone on the box